### PR TITLE
[Core] Remove lookup remote connections and create object manager client pool instead

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Currently, we have a weird API within object directory that looks up all connections. This will have hard-to-read code path within pull manager refactoring (since we need to path a callback that uses obod method, which is weird). To fix it, I wrote a object_manager_client_pool, which can look up alive nodes and create a connection 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/object_manager_client_pool.h
Comment (rkooo567) [inline src/ray/object_manager/object_manager_client_pool.h:44]: I think this method could be private actually

Comment (rkooo567) [reply src/ray/object_manager/object_manager_client_pool.h:44]: Looks like making this public cause deadlock. Let me fix it.

**Comment by rkooo567** [comment]
@scv119 Lmk if you think we need unit tests here

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:25]: nits: [don't use namespace in header](https://docs.microsoft.com/en-us/cpp/cpp/namespaces-cpp?view=msvc-160#:~:text=In%20general%2C%20avoid%20putting%20using,names%20in%20a%20header%20file.)

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:45]: do we really need Disconnect? i.e. seems not used.

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:91]: could it be a reference?

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:39]: > The returned pointer is borrowed, and expected to be used briefly.

this looks overkill? at least this is inconsistent with the API. shared_ptr means shared ownership so it's not borrowed but rather hold a shared ownership.

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:64]: ah i don't think you need define this.

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:52]: some comments on thread safety?

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:49]: GetOrConnect... ?

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:30]: nits: this could be defined inside ObjectManagerClientPool

File: src/ray/object_manager/object_manager_client_pool.h
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.h:73]: super nits: prefer explicit capture (&ccm) in this case.

File: src/ray/object_manager/object_manager_client_pool.cc
Comment (scv119) [inline src/ray/object_manager/object_manager_client_pool.cc:36]: this involves an io even if the id is cached. what happens if we always goes to client_map first?

**Review by scv119 — COMMENTED** [review]
round 1

**Comment by rkooo567** [comment]
Let me address this tomorrow. Btw, most of code is copied from node_manager_client_pool lol

File: src/ray/object_manager/object_manager_client_pool.h
Comment (mwtian) [reply src/ray/object_manager/object_manager_client_pool.h:39]: Is the `absl::optional<>` wrapper necessary to represent an error condition, or a non-empty `absl::optional<>` always contains a non-empty `std::shared_ptr` here?

**Review by scv119 — APPROVED** [review]

**Comment by rkooo567** [comment]
The unit test requires some mocking on gcs client, which is already done in OBOD unit test in https://github.com/ray-project/ray/pull/18016. So I will merge this PR after we merge that PR

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**
